### PR TITLE
🌱 Detect certificate expiry from kube-apiserver serving cert

### LINKED ISSUE #7342: KCP should introspect machine certificate expiry information from node to enable automatic certificate renewal
**User Story**

Recently a feature was introduced in KCP to enable automatic renewal of certificates by doing a repave: https://github.com/kubernetes-sigs/cluster-api/pull/6983

This feature only works for control plane machines created using Cluster API v1.3 or above. 
For older control plane machines it needs manual intervention from the user to provide the actual certificate expiry date of the control plane machines. 

This issue is to explore ideas around how we can introspect the certificate expiry time from the node without needing manual intervention from the user.

**Detailed Description**

The certificate expiry time cannot reliably calculated from the `creationTimestamp` of the Machine object. Example: If the machine object is restored from a backup the `creationTimestamp` would have changed. 

 One possible way to calculate the certificate expiry time:
 > Cluster API / KCP introspect these values by running an exec pod temporarily on the node
>
> xref: https://github.com/kubernetes-sigs/cluster-api/pull/7268#discussion_r984800820

/kind feature
/area control-plane

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**WIP**: waiting for feedback before implementing unit tests

This PR changes the detection of the certificate expiry. Instead of setting it only for new machines 
in CABPK now we detect it in KCP from the serving certificate of the kube-apiserver.

This has the big advantage that it allows detecting certificate expiry of pre-existing machines!

For more details see: https://github.com/kubernetes-sigs/cluster-api/issues/7342#issuecomment-1267506083

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7342

### CODE REVIEW

**Comment by sbueringer** [comment]
/assign @fabriziopandini @vincepri @ykakarap 

Please take a look. Quick feedback would be really appreciated!

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (neolit123) [inline controlplane/kubeadm/internal/controllers/controller.go:626]: should this log line include node name?

**Review by neolit123 — COMMENTED** [review]
i think this approach is ok

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:626]: Thx for the review and yes good catch!

File: controlplane/kubeadm/internal/workload_cluster.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster.go:445]: Good idea. I'll take a look how kubeadm generates the kube-apiserver static pod

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:614]: I had it here because the map lookup on the nil map looked dangerous. But it actually works... so moved it down

**Comment by sbueringer** [comment]
@fabriziopandini Should be both adressed 

/test pull-cluster-api-e2e-full-main

File: controlplane/kubeadm/internal/workload_cluster.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster.go]: You're right, fixed!

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:624]: Fixed

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

File: controlplane/kubeadm/internal/workload_cluster.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster.go:496]: Good point...

I was just happy that I could handle this case relatively easily

**Comment by sbueringer** [comment]
/hold

Waiting for feedback from Vince and then I want to add some test coverage before merge.

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (vincepri) [inline controlplane/kubeadm/internal/controllers/controller.go:622]: Should we also add the machine name if it doesn't have it already?

**Review by vincepri — APPROVED** [review]
/approve

File: controlplane/kubeadm/internal/controllers/controller.go
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:622]: Should be covered by l.603
> 		log = log.WithValues("Machine", klog.KObj(m))

**Comment by vincepri** [comment]
Ready to unhold/unwip?

**Comment by sbueringer** [comment]
I would add unit tests tomorrow and Fabrizio can lgtm afterwards

**Review by ykakarap — COMMENTED** [review]
The overall approach looks good 👍🏼 
This will be a nice improvements for the users.

/lgtm

**Comment by sbueringer** [comment]
Added unit test, now ready for merge

**Review by fabriziopandini — COMMENTED** [review]
Great work!
/lgtm
/approve

**Pull request merged**
